auraed/src/ebpf/tracepoint_programs/tracepoint_program.rs

```diff
@@ -0,0 +1,130 @@
+        // the event
+        let per_cpu_buffer_capacity = (PER_CPU_BUFFER_SIZE_IN_PAGES
+            * page_size as usize)
+            / event_struct_size;
+
+        // Set the capacity of the channel to the combined capacity of all the
+        // per-CPU buffers
+        let channel_capacity = per_cpu_buffer_capacity * num_cpus;
+
+        // Attach to kernel trace event
+        let _ = program.attach(Self::CATEGORY, Self::EVENT)?;
+
+        // Create the channel for broadcasting the events
+        let (tx, _) = broadcast::channel(channel_capacity);
+
+        // Open the BPF_PERF_EVENT_ARRAY BPF map that is used to send data from
+        // kernel to userspace. This array contains the per-CPU buffers and is
+        // indexed by CPU id.
+        // https://libbpf.readthedocs.io/en/latest/api.html
+        let mut perf_array =
+            AsyncPerfEventArray::try_from(bpf.map_mut(Self::PERF_BUFFER)?)?;
+
+        // Spawn a thread per CPU to listen for events from the kernel.
+        for cpu_id in online_cpus()? {
+            trace!("spawning task for cpu {cpu_id}");
+            // Open the per-CPU buffer for the current CPU id
+            let mut per_cpu_buffer =
+                perf_array.open(cpu_id, Some(PER_CPU_BUFFER_SIZE_IN_PAGES))?;
+
+            // Clone the sender of the event broadcast channel
+            let per_cpu_tx = tx.clone();
+
+            // Spawn the thread to listen on the per-CPU buffer
+            let _ignored = tokio::spawn(async move {
+                trace!("task for cpu {cpu_id} awaiting for events");
+
+                // Allocate enough memory to drain the entire buffer
+                // Note: using `vec!` macro will not result in a correct `Vec`
+                let mut buffers = (0..per_cpu_buffer_capacity)
+                    .map(|_| BytesMut::with_capacity(event_struct_size))
+                    .collect::<Vec<_>>();
+
+                // Start polling the per-CPU buffer for events
+                loop {
+                    let events = match per_cpu_buffer
+                        .read_events(&mut buffers)
+                        .await
+                    {
+                        Ok(events) => events,
+                        Err(error) => {
+                            error!("fail to read events from per-cpu perf buffer, bailing out: {error}");
+                            return;
+                        }
+                    };
+
+                    if events.lost > 0 {
+                        error!(
+                            "buffer full, dropped {} perf events - this should never happen!",
+                            events.lost
+                        );
+                    }
+
+                    for buf in buffers.iter_mut().take(events.read) {
+                        let ptr = buf.as_ptr() as *const T;
+                        let signal = unsafe { ptr.read_unaligned() };
+                        if per_cpu_tx.send(signal).is_err() {
```

Comment:
If there are no listeners, I think it's safe to ignore the error without having the `error!` trace.

Reply:
When would this be the case though? 🤔 

FWIW with future tracepoints we are going to have to be a bit more mindful wrt resource consumption. This listener is continuously listening for signals, which is fine for now. But if we're instrumenting all syscalls on a machine I think we should consider just running that probe in the context of the gRPC request. Or maybe share a BPF map between kernel and userspace that has some flags which "features" we want enabled which the probe then checks before running any instrumentation code.

Reply:
As far as I see, having no listeners is the default case. The only time we have a listener is when a client subscribes via `ObserveService::get_posix_signals_stream`

Reply:
Ah ok, I thought `send()` would only error if the channel was closed, and didn't care if there were receivers or not. Wouldn't we be seeing tons of error messages otherwise when running the daemon?

Related: maybe `PerfEventListener` is confusing, as it's really a wrapper around a `Sender<T>`.

Reply:
I think LogChannel is the only other place we use broadcast channels and we also ignore the error for that, so we don't have any tracing of these errors.

RE: `PerfEventListener`...I assumed there would be a struct that could be subscribed, but wouldn't expose send in tokio, but didn't find one, so I actually made another struct (which I called `Subscribable`, because names are hard) before realizing you did the same in `PerfEventListener`, so I think it's the right decision to have it.

Reply:
Hmm, maybe 'PerfEventBroadcast`...I think I'll update to that if there are no objections.